### ISSUE TITLE
API requests error

### ISSUE THREAD

**Issue by brokerelcom** [gradio-app/gradio#10710]
### Describe the bug

Hello! 
I am using SDK Gradio on on my Public space on Huggingface.  I am deployed AI chat bot.  WEB works OK. All requests and responses via terminal Bash  - OK. But when I try to connect my telegram-bot to Gradio API   - then i have error in log: server rejected WebSocket connection: HTTP 403. 
Could you please advice me what is wrong with requests via telegram?
Thanks!

On Hugging face I am using:
 gradio_client==0.6.1
huggingface-hub>=0.19.3,<1.0
gradio==4.26.0

All API requests are done according to Gradio:

api_name: /chat
from gradio_client import Client
client = Client("dkmanika/nika_prop")
result = client.predict(
		message="Hello!!",
		api_name="/chat"
)
print(result)

### Have you searched existing issues?  🔎

- [x] I have searched and found no existing issues

### Reproduction

```python

import os
import asyncio
import logging
import json
import random
import psutil
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import (
    ApplicationBuilder,
    CommandHandler,
    MessageHandler,
    CallbackQueryHandler,
    filters,
    CallbackContext,
)
from dotenv import load_dotenv
from gradio_client import Client


```
#request to Gradio API, telegram-bot code, part, Python 3.19.  

async def send_request_to_gradio(query: str, state: str = None) -> str:

    try:
        await asyncio.sleep(random.uniform(1.0, 2.0))
        client = Client(HF_SPACE_NAME, hf_token=HF_TOKEN)

        result = client.predict([
            query, state
        ], api_name="/chat")

        response = result[0] if isinstance(result, list) and result else "Не найдено"
        return response
    except Exception as e:
        logging.error(f"ERROR requests to Gradio: {e}")
        return "ERROR requests to Gradio. Try again."

### Logs

```shell
===== Application Startup at 2025-02-28 12:28:46 =====


tokenizer_config.json:   0%|          | 0.00/453 [00:00<?, ?B/s]
tokenizer_config.json: 100%|██████████| 453/453 [00:00<00:00, 3.56MB/s]

tokenizer.json:   0%|          | 0.00/16.3M [00:00<?, ?B/s]
tokenizer.json: 100%|██████████| 16.3M/16.3M [00:00<00:00, 148MB/s]

added_tokens.json:   0%|          | 0.00/23.0 [00:00<?, ?B/s]
added_tokens.json: 100%|██████████| 23.0/23.0 [00:00<00:00, 197kB/s]

special_tokens_map.json:   0%|          | 0.00/173 [00:00<?, ?B/s]
special_tokens_map.json: 100%|██████████| 173/173 [00:00<00:00, 1.61MB/s]

config.json:   0%|          | 0.00/879 [00:00<?, ?B/s]
config.json: 100%|██████████| 879/879 [00:00<00:00, 8.16MB/s]

model.safetensors:   0%|          | 0.00/1.11G [00:00<?, ?B/s]
model.safetensors:   3%|▎         | 32.9M/1.11G [00:01<00:40, 26.5MB/s]
model.safetensors:  43%|████▎     | 473M/1.11G [00:02<00:02, 244MB/s]  
model.safetensors: 100%|█████████▉| 1.11G/1.11G [00:02<00:00, 380MB/s]
/usr/local/lib/python3.10/site-packages/gradio/components/chatbot.py:228: UserWarning: The 'tuples' format for chatbot messages is deprecated and will be removed in a future version of Gradio. Please set type='messages' instead, which uses openai-style 'role' and 'content' keys.
  warnings.warn(
* Running on local URL:  http://0.0.0.0:7860, with SSR ⚡

To create a public link, set `share=True` in `launch()`.
```

### System Info

```shell
I cant find the reason of error
```

### Severity

Blocking usage of gradio

**Comment by abidlabs** [comment]
Hi @brokerelcom can you please try with the latest version of Gradio? (5.20).

We do not maintain Gradio 4.x and cannot issue backfixes. See also: https://github.com/gradio-app/gradio/issues/9463

**Comment by brokerelcom** [comment]
Hello and thanks for reply!

Not yet, only the latest 4.x.x.

When I tried 5.19 I got a lot of Errors in many  of the dependencies.

Could you please provide me info how to upgrade other dependencies to be
able work with 5.19 / 5.20?

I have the list:

huggingface-hub>=0.19.3,<1.0
transformers>=4.35.0
tokenizers>=0.15.0,<1.0
torch>=1.13.1
pymorphy2>=0.8
regex>=2023.10.3
pandas>=2.1.0
gradio==4.44.1
gradio-client==1.3.0
urllib3~=2.0
requests>=2.28.2
python-telegram-bot>=20.2
httpx>=0.24.1
aiohttp>=3.8.5,<4.0.0
async-timeout==4.0.2
python-dotenv==1.0.0


Thanks!!!

**Comment by abidlabs** [comment]
I think you should just need to not pin `httpx`, `aiohttp`, `huggingface-hub` and `gradio-client`

**Comment by abidlabs** [comment]
Feel free to ask follow up questions like this (that are not feature requests or bug reports) in [our Discord server](http://discord.gg/feTf9x3ZSB). (I'll close this issue)

**Comment by brokerelcom** [comment]
updated to Gradio 5.19 but still "ERROR in connection to Gradio: server rejected WebSocket connection: HTTP 403"
Here is requirements file for telegram bot and for HF-bot. 

TG bot requitements:

psutil 
huggingface-hub>=0.19.3
openpyxl==3.1.2
pandas>=2.1.0 
transformers==4.35.0 
torch>=1.13.1
tokenizers>=0.15.0,<1.0
pymorphy2>=0.8 
regex>=2023.10.3 
python-telegram-bot==20.2 
gradio-client>=1.3.0.
python-dotenv==1.0.0       
requests>=2.28.2
urllib3~=2.0 
python-telegram-bot>=20.2
httpx>=0.24.1 
aiohttp>=3.8.5 
async-timeout==4.0.2


HF bot requitements:

huggingface-hub>=0.19.3 
transformers>=4.35.0
tokenizers>=0.15.0,<1.0
torch>=1.13.1
pymorphy2>=0.8
regex>=2023.10.3
pandas>=2.1.0 
gradio==5.19.0
gradio-client>=1.3.0
urllib3~=2.0
requests>=2.28.2
python-telegram-bot>=20.2  # Для работы с Telegram API
httpx>=0.24.1

**Comment by abidlabs** [comment]
Could you put together a minimal repro for us to look at? For both the Gradio app itself and the Gradio Client code? Thanks. https://stackoverflow.com/help/minimal-reproducible-example

**Comment by brokerelcom** [comment]
Telegram bot:

import os
import asyncio
import logging
import json
import random
import psutil
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import (
    ApplicationBuilder,
    CommandHandler,
    MessageHandler,
    CallbackQueryHandler,
    filters,
    CallbackContext,
)
from dotenv import load_dotenv
from gradio_client import Client

load_dotenv()


TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN")
HF_SPACE_NAME = os.getenv("HF_SPACE_NAME")  
HF_TOKEN = os.getenv("HF_TOKEN")

///main code body///


# Request API Gradio
async def send_request_to_gradio(query: str, state: str = None) -> str:

    try:
        await asyncio.sleep(random.uniform(1.0, 2.0))
        client = Client(HF_SPACE_NAME, hf_token=HF_TOKEN)

        result = client.predict([
            query, state
        ], api_name="/chat")

        response = result[0] if isinstance(result, list) and result else "Не найдено"
        return response
    except Exception as e:
        logging.error(f"ERROR request to Gradio: {e}")
        return "ERROR request to Gradio. Try again."


async def handle_message(update: Update, context: CallbackContext) -> None:
    user_input = update.message.text
    state = context.user_data.get("state", None)
    model_answer = await send_request_to_gradio(user_input, state)
    await asyncio.sleep(random.uniform(1.5, 3.0))
    await update.message.reply_text(model_answer)
    context.user_data["state"] = model_answer
    log_dialog(user_input, model_answer)


def start_hf_prop_bot() -> None:
    application = ApplicationBuilder().token(TELEGRAM_BOT_TOKEN).build()
    application.add_handler(CommandHandler("start", start))
    application.add_handler(CallbackQueryHandler(stop_bot, pattern="^stop_bot$"))
    application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_message))
    application.run_polling()
    logging.info("Bot started.")

if __name__ == "__main__":
    start_hf_prop_bot()


HF bot:

import gradio as gr
import json
import pandas as pd
from transformers import AutoTokenizer, AutoModelForQuestionAnswering
import torch
import re
from pymorphy2 import MorphAnalyzer
import logging

# logger
logging.basicConfig(
    filename="dialog_logs.log", 
    level=logging.INFO,         
    format="%(asctime)s - %(message)s",  
    datefmt="%Y-%m-%d %H:%M:%S"  
)

# dialog log
def log_dialog(user_input, model_answer):
    if user_input.strip() == "" or model_answer.strip() == "":
        return  # Не записываем пустые диалоги
    
    logging.info(f"User: {user_input}")
    logging.info(f"Model: {model_answer}")


# model load
model_name = "timpal0l/mdeberta-v3-base-squad2"
tokenizer = AutoTokenizer.from_pretrained(model_name, timeout=35)
model = AutoModelForQuestionAnswering.from_pretrained(model_name)


# JSON
def load_questions_from_json(file_path="questions.json"):
    with open(file_path, "r", encoding="utf-8") as file:
        return json.load(file)


# search JSON
def find_answer_in_json(questions_data, user_message):
    for item in questions_data:
        if user_message.lower() in item["question"].lower():
            return item["answer"]
    return None


# load data CSV
def load_real_estate_data(file_path="real_estate.csv"):
    try:
        data = pd.read_csv(file_path, sep=";", encoding="utf-8")
        if data.empty:
            raise ValueError("Empty or corrupted")
        return data
    except Exception as e:
        print(f"Ошибка при загрузке файла данных: {e}")
        return pd.DataFrame(columns=["item_id", "address", "metro", "met_range", "price", "description", "type", "area", "RENT"])


morph = MorphAnalyzer()
base_types = ["офис", "квартира", "апартаменты", "свободное назначение", "студия", "дом", "ОСЗ"]



# user rquest analyzer
def analyze_query(query):
    query = query.lower()
    filters = {}
    words = query.split()
    lemmas = [morph.parse(word)[0].normal_form for word in words]

    # search
    types_found = []
    for base_type in base_types:
        if any(morph.parse(lemma)[0].normal_form == base_type for lemma in lemmas):
            types_found.append(base_type)
    if types_found:
        filters["type"] = types_found

    # dea search
    if any(word in query for word in ["аренда", "снять", "аренду", "арендовать", "сниму"]):
        filters["rent"] = "Аренда"
    elif any(word in query for word in ["продажа", "купить", "покупке", "покупки", "продаю", "продажи"]):
        filters["rent"] = "Продажа"

    
    def normalize_number(text):
        match = re.search(r"(\d+[\s]*[.,]?\d*)\s*(млн|тр|т\.р|тыс|тысруб|тыс\.руб|тр|т\.р.|тыр|тыр.|тыщ|тыш|тысяч|млнруб|млн\.руб|мн|М|миллионов|милионов|лямов|лимонов)?", text)
        if not match:
            return None
        number = float(match.group(1).replace(" ", "").replace(",", "."))
        unit = match.group(2)
        if unit in ["млн", "м", "млнруб", "млн.руб", "мн", "М", "миллионов", "милионов", "лямов", "лимонов"]:
            return int(number * 1_000_000)
        elif unit in ["тр", "т.р", "т.р.", "тыс", "тысруб", "тыс.руб", "тыр", "тыр.", "тыщ", "тыш", "тысяч"]:
            return int(number * 1_000)
        else:
            return int(number)

    # price search
    price_keywords = ["руб", "р.", "рублей", "рубля"]
    price_range_match = re.findall(rf"от\s*(\d+[\s]*[.,]?\d*)\s*до\s*(\d+[\s]*[.,]?\d*)\s*({'|'.join(price_keywords)})\b", query)
    if price_range_match:
        raw_price_min = price_range_match[0][0]
        raw_price_max = price_range_match[0][1]
        normalized_price_min = normalize_number(raw_price_min)
        normalized_price_max = normalize_number(raw_price_max)
        if normalized_price_min is not None and normalized_price_max is not None:
            filters["price_min"] = normalized_price_min
            filters["price_max"] = normalized_price_max

  
    price_match = re.findall(rf"(до|\s*)(\d+[\s]*[.,]?\d*)\s*(млн|тр|т\.р|тыс|тысруб|тыс\.руб|тр|т\.р.|тыр|тыр.|тыщ|тыш|тысяч|млнруб|млн\.руб|мн|М|миллионов|милионов|лямов|лимонов)?\s*({'|'.join(price_keywords)})\b", query)
    if price_match:
        operator = price_match[0][0]
        raw_price = price_match[0][1] + (price_match[0][2] or "")
        normalized_price = normalize_number(raw_price)
        if normalized_price is not None:
            if operator in ["до", " "]:
                filters["price_max"] = normalized_price
            else:
                filters["price_min"] = normalized_price

    # area search
    area_keywords = ["квм", "кв.м", "кв метров", "кв м", "кв.м.", "квадратов", "квадрата", "м2", "метров", "м²"]
    area_range_match = re.findall(rf"от\s*(\d+[\s]*[.,]?\d*)\s*до\s*(\d+[\s]*[.,]?\d*)\s*({'|'.join(area_keywords)})\b", query)
    if area_range_match:
        raw_area_min = area_range_match[0][0]
        raw_area_max = area_range_match[0][1]
        normalized_area_min = normalize_number(raw_area_min)
        normalized_area_max = normalize_number(raw_area_max)
        if normalized_area_min is not None and normalized_area_max is not None:
            filters["area_min"] = normalized_area_min
            filters["area_max"] = normalized_area_max


    area_min_match = re.findall(rf"\b(от)\s*(\d+)\s*({'|'.join(area_keywords)})", query)
    area_max_match = re.findall(rf"\b(до)\s*(\d+)\s*({'|'.join(area_keywords)})", query)
    if area_min_match:
        filters["area_min"] = int(area_min_match[0][1])
    if area_max_match:
        filters["area_max"] = int(area_max_match[0][1])

    return filters


# data search in  CSV
def search_real_estate(dataframe, filters):
    if not filters:
        return []

    mask = pd.Series([True] * len(dataframe))

    # filter 1
    if "type" in filters:
        mask &= dataframe["type"].str.lower().isin(filters["type"])

    # Фильтрация по виду сделки (Аренда/Продажа)
    if "rent" in filters:
        mask &= dataframe["RENT"].str.lower() == filters["rent"].lower()

    # price filter
    if "price_min" in filters and "price_max" in filters:
        mask &= (dataframe["price"] >= filters["price_min"]) & (dataframe["price"] <= filters["price_max"])
    elif "price_min" in filters:
        mask &= dataframe["price"] >= filters["price_min"]
    elif "price_max" in filters:
        mask &= dataframe["price"] <= filters["price_max"]

    # area filter
    if "area_min" in filters and "area_max" in filters:
        mask &= (dataframe["area"] >= filters["area_min"]) & (dataframe["area"] <= filters["area_max"])
    elif "area_min" in filters:
        mask &= dataframe["area"] >= filters["area_min"]
    elif "area_max" in filters:
        mask &= dataframe["area"] <= filters["area_max"]

  
    return dataframe[mask].to_dict(orient="records")

# context for model
def create_context(dataframe, filters):
    if "type" in filters and "rent" in filters:
        
        relevant_data = dataframe[
            (dataframe["type"].str.lower().isin(filters["type"])) &
            (dataframe["RENT"].str.lower() == filters["rent"].lower())
        ]
    else:
        
        relevant_data = dataframe
    context = " ".join(relevant_data["description"].dropna().tolist()) + " " + \
              " ".join(relevant_data["type"].dropna().tolist()) + " " + \
              " ".join(relevant_data["address"].dropna().tolist())
    return context


# answer from model
def get_model_response(question, context=""):
    inputs = tokenizer.encode_plus(question, context, return_tensors="pt", max_length=512, truncation=True)
    outputs = model(**inputs)
    answer_start = torch.argmax(outputs.start_logits)
    answer_end = torch.argmax(outputs.end_logits) + 1

    
    if answer_start >= answer_end:
        return "Could not find"

    answer = tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(inputs["input_ids"][0][answer_start:answer_end]))
    return answer.strip() if answer.strip() != "[CLS]" else "Not found"

    


# main Gradio
def predict(message, history=None):
    if history is None:
        history = []

    # load JSON/CVS
    questions_data = load_questions_from_json()
    real_estate_data = load_real_estate_data()

    #search in JSON
    json_answer = find_answer_in_json(questions_data, message)
    if json_answer:
        log_dialog(message, json_answer)
        return json_answer

    # analyze
    filters = analyze_query(message)

    # search
    filtered_results = search_real_estate(real_estate_data, filters)
    if filtered_results:
        response = ""
        for result in filtered_results:
            response += (
                f"ID: {result.get('item_id', 'Не указано')}\n"
                f"Тип: {result.get('type', 'Не указано')}\n"
                f"Адрес: {result.get('address', 'Не указано')}\n"
                f"Цена: {result.get('price', 'Не указано')} руб.\n"
                f"Площадь: {result.get('area', 'Не указано')} м²\n"
                f"Метро: {result.get('metro', 'Не указано')}\n"
                f"До метро: {result.get('met_range', 'Не указано')}\n"
                f"Описание: {result.get('description', 'Не указано')}\n"
                f"Фото: {result.get('foto', 'Не указано')}\n"
                f"Записаться на просмотр: {result.get('order', 'Не указано')}\n"
                f"---\n"
            )
        log_dialog(message, response)
        return response

    
    context = " ".join(real_estate_data["description"].tolist())
    model_answer = get_model_response(message, context)
    log_dialog(message, model_answer)
    return model_answer


# interface Gradio
demo = gr.ChatInterface(
    fn=predict,
    title="AI assistamt",
    description="Ask and I will find",
)

if __name__ == "__main__":
    demo.launch()
